state/accountsDB.go

```diff
@@ -1089,53 +1105,89 @@ func (adb *AccountsDB) SnapshotState(rootHash []byte) {
+		trieStorageManager.ExitPruningBufferingMode()
+
+		stats.wg.Done()
+	}()
+
+	go adb.markActiveDBAfterSnapshot(stats, errChan, rootHash, "snapshotState user trie", epoch)
+
+	adb.waitForCompletionIfRunningInImportDB(stats)
+}
+
+func (adb *AccountsDB) prepareSnapshot(rootHash []byte) (common.StorageManager, uint32, bool) {
+	trieStorageManager, epoch, err := adb.getTrieStorageManagerAndLatestEpoch()
+	if err != nil {
+		log.Error("snapshot user state error", "err", err.Error())
+		return nil, 0, false
+	}
+
+	if !adb.shouldTakeSnapshot(trieStorageManager, rootHash, epoch) {
 		log.Debug("skipping snapshot",
 			"last snapshot rootHash", adb.lastSnapshot.rootHash,
 			"rootHash", rootHash,
 			"last snapshot epoch", adb.lastSnapshot.epoch,
 			"epoch", epoch,
+			"isSnapshotInProgress", adb.isSnapshotInProgress.IsSet(),
 		)
-		return
+		return nil, 0, false
 	}
 
-	log.Info("starting snapshot", "rootHash", rootHash, "epoch", epoch)
-
+	adb.isSnapshotInProgress.SetValue(true)
 	adb.lastSnapshot.rootHash = rootHash
 	adb.lastSnapshot.epoch = epoch
 	err = trieStorageManager.Put([]byte(lastSnapshotStarted), rootHash)
 	handleLoggingWhenError("could not set lastSnapshotStarted", err, "rootHash", rootHash)
-
 	trieStorageManager.EnterPruningBufferingMode()
 
-	errChan := make(chan error, 1)
-	stats := newSnapshotStatistics(1)
-	go func() {
-		leavesChannel := make(chan core.KeyValueHolder, leavesChannelSize)
-		stats.NewSnapshotStarted()
-		trieStorageManager.TakeSnapshot(rootHash, rootHash, leavesChannel, errChan, stats, epoch)
-		adb.snapshotUserAccountDataTrie(true, rootHash, leavesChannel, errChan, stats, epoch)
-		trieStorageManager.ExitPruningBufferingMode()
+	return trieStorageManager, epoch, true
+}
 
-		stats.wg.Done()
-	}()
+func (adb *AccountsDB) getTrieStorageManagerAndLatestEpoch() (common.StorageManager, uint32, error) {
+	trieStorageManager := adb.mainTrie.GetStorageManager()
+	epoch, err := trieStorageManager.GetLatestStorageEpoch()
+	if err != nil {
+		return nil, 0, fmt.Errorf("%w while getting the latest storage epoch", err)
+	}
 
-	go adb.markActiveDBAfterSnapshot(stats, errChan, rootHash, "snapshotState user trie", epoch)
+	return trieStorageManager, epoch, nil
+}
 
-	adb.waitForCompletionIfRunningInImportDB(stats)
+func (adb *AccountsDB) shouldTakeSnapshot(trieStorageManager common.StorageManager, rootHash []byte, epoch uint32) bool {
+	snapshotAlreadyTaken := bytes.Equal(adb.lastSnapshot.rootHash, rootHash) && adb.lastSnapshot.epoch == epoch
+	if snapshotAlreadyTaken {
+		return false
+	}
+
+	if adb.isSnapshotInProgress.IsSet() {
+		return false
+	}
+
+	if !trieStorageManager.ShouldTakeSnapshot() {
+		return false
+	}
+
+	return true
 }
 
 func (adb *AccountsDB) markActiveDBAfterSnapshot(stats *snapshotStatistics, errChan chan error, rootHash []byte, message string, epoch uint32) {
 	stats.PrintStats(message, rootHash)
 
+	defer adb.isSnapshotInProgress.Reset()
```

Comment:
So the function `markActiveDBAfterSnapshot()` is now more like a `doStuffOnSnapshotCompletion`.

Reply:
Right. Renamed to `processSnapshotCompletion`. I look forward to a better naming idea :sweat_smile: